op-e2e/actions/user.go

```diff
@@ -0,0 +1,322 @@
+		GasTipCap: s.txOpts.GasTipCap,
+		Value:     s.TxValue(),
+		ChainID:   s.env.Signer.ChainID(),
+		Nonce:     s.PendingNonce(t),
+		Gas:       gas,
+	})
+	err = s.env.EthCl.SendTransaction(t.Ctx(), tx)
+	require.NoError(t, err, "must send tx")
+	s.lastTxHash = tx.Hash()
+}
+
+func (s *BasicUser[B]) ActCheckReceiptStatusOfLastTx(success bool) func(t Testing) {
+	return func(t Testing) {
+		s.CheckReceipt(t, success, s.lastTxHash)
+	}
+}
+
+func (s *BasicUser[B]) CheckReceipt(t Testing, success bool, txHash common.Hash) *types.Receipt {
+	receipt, err := s.env.EthCl.TransactionReceipt(t.Ctx(), txHash)
+	if receipt != nil && err == nil {
+		expected := types.ReceiptStatusFailed
+		if success {
+			expected = types.ReceiptStatusSuccessful
+		}
+		require.Equal(t, expected, receipt.Status, "expected receipt status to match")
+		return receipt
+	} else if err != nil && !errors.Is(err, ethereum.NotFound) {
+		t.Fatalf("receipt for tx %s was not found", txHash)
+	} else {
+		t.Fatalf("receipt error: %v", err)
+	}
+	return nil
+}
+
+type L1User struct {
+	BasicUser[*L1Bindings]
+}
+
+type L2User struct {
+	BasicUser[*L2Bindings]
+}
+
+type CrossLayerUser struct {
+	L1 L1User
+	L2 L2User
+
+	lastL1DepositTxHash common.Hash
+}
+
+func NewCrossLayerUser(log log.Logger, priv *ecdsa.PrivateKey, rng *rand.Rand) *CrossLayerUser {
+	addr := crypto.PubkeyToAddress(priv.PublicKey)
+	return &CrossLayerUser{
+		L1: L1User{
+			BasicUser: BasicUser[*L1Bindings]{
+				log:     log,
+				rng:     rng,
+				account: priv,
+				address: addr,
+			},
+		},
+		L2: L2User{
+			BasicUser: BasicUser[*L2Bindings]{
+				log:     log,
+				rng:     rng,
+				account: priv,
+				address: addr,
+			},
+		},
+	}
+}
+
+func (s *CrossLayerUser) ActDeposit(t Testing) {
+	isCreation := false
+	toAddr := common.Address{}
+	if s.L2.txToAddr == nil {
+		isCreation = true
+	} else {
+		toAddr = *s.L2.txToAddr
+	}
+	depositTransferValue := s.L2.TxValue()
+	depositGas := s.L2.txOpts.GasLimit
+	if s.L2.txOpts.GasLimit == 0 {
+		// estimate gas used by deposit
+		gas, err := s.L2.env.EthCl.EstimateGas(t.Ctx(), ethereum.CallMsg{
+			From:       s.L2.address,
+			To:         s.L2.txToAddr,
+			Value:      depositTransferValue, // TODO: estimate gas does not support minting yet
+			Data:       s.L2.txCallData,
+			AccessList: nil,
+		})
+		require.NoError(t, err)
+		depositGas = gas
+	}
+
+	tx, err := s.L1.env.Bindings.OptimismPortal.DepositTransaction(&s.L1.txOpts, toAddr, depositTransferValue, depositGas, isCreation, s.L2.txCallData)
+	require.NoError(t, err, "failed to create deposit tx")
+
+	// Send the actual tx (since tx opts don't send by default)
+	err = s.L1.env.EthCl.SendTransaction(t.Ctx(), tx)
+	require.NoError(t, err, "must send tx")
+	s.lastL1DepositTxHash = tx.Hash()
+}
+
+func (s *CrossLayerUser) ActCheckDepositStatus(l1Success, l2Success bool) Action {
+	return func(t Testing) {
+		s.CheckDepositTx(t, s.lastL1DepositTxHash, 0, l1Success, l2Success)
```

Comment:
Hardcoding 0 makes this less flexible but simple

Reply:
Exactly, this is a good example of when the higher level action (checking a deposit) omits some lower level edge case (checking tx success). When we do need the lower level detail in a test, we can use the underlying lower-level actions.